File: client/src/utils/chart/blendFade2Signals.ts
import type { ChartCandle } from '@/types/candle'
import {
  DEFAULT_BLEND_FADE2_CONFIG,
  type BlendFade2Config,
} from '@/types/blendFade2Config'
import type { MarketTrend } from '@/utils/chart/detectBreakOfStructure'

export interface BlendFade2SignalArrays {
  entryBar: boolean[]
  side: (MarketTrend | null)[]
}

function sessionOk(openTimeUnix: number, cfg: BlendFade2Config): boolean {
  const start = cfg.sessionUtcStart
  const end = cfg.sessionUtcEnd
  if (start === null || end === null) return true
  const hour =
    openTimeUnix > 1_000_000_000_000
      ? Math.floor((openTimeUnix / 1000 / 3600) % 24)
      : Math.floor((openTimeUnix / 3600) % 24)
  if (start <= end) return hour >= start && hour < end
  return hour >= start || hour < end
}

function zScore(values: number[], endIdx: number, lookback: number): number | null {
  const start = endIdx - lookback
  if (start < 0) return null
  let sum = 0
  for (let i = start; i < endIdx; i++) sum += values[i]
  const mu = sum / lookback
  let varSum = 0
  for (let i = start; i < endIdx; i++) {
    const d = values[i] - mu
    varSum += d * d
  }
  const std = Math.sqrt(varSum / lookback)
  if (std <= 0) return null
  return (values[endIdx] - mu) / std
}

function percentileRank(closes: number[], endIdx: number, lookback: number): number | null {
  const start = endIdx - lookback
  if (start < 0) return null
  let lo = closes[start]
  let hi = closes[start]
  for (let i = start; i <= endIdx; i++) {
    lo = Math.min(lo, closes[i])
    hi = Math.max(hi, closes[i])
  }
  const span = hi - lo
  if (span <= 0) return null
  return (closes[endIdx] - lo) / span
}

/** 1:1 with trading-cursor-models strategies/blend_fade2/signals.py */
export function generateBlendFade2Signals(
  candles: ChartCandle[],
  config: BlendFade2Config = DEFAULT_BLEND_FADE2_CONFIG,
): BlendFade2SignalArrays {
  const n = candles.length
  const entryBar: boolean[] = Array(n).fill(false)
  const side: (MarketTrend | null)[] = Array(n).fill(null)

  const closes = candles.map((c) => c.close)
  const openTimes = candles.map((c) => c.time)

  const lb = config.lookback
  const lbF = config.lookbackFast
  const zTh = config.zThreshold

  for (let i = 1; i < n; i++) {
    if (!sessionOk(openTimes[i], config)) continue

    const closed = i - 1
    if (closed < Math.max(lb, lbF) + 1) continue

    const z = zScore(closes, closed, lb)
    if (z === null) continue

    if (config.minRangePct > 0 && closed >= lb) {
      const windowStart = closed - lb
      const ref = closes[windowStart]
      if (ref > 0) {
        let windowMax = closes[windowStart]
        let windowMin = closes[windowStart]
        for (let j = windowStart; j <= closed; j++) {
          windowMax = Math.max(windowMax, closes[j])
          windowMin = Math.min(windowMin, closes[j])
        }
        const move = (windowMax - windowMin) / ref
        if (move < config.minRangePct) continue
      }
    }

    let signalSide: 'long' | 'short' | null = null
    if (z > zTh) signalSide = 'short'
    else if (z < -zTh) signalSide = 'long'
    if (signalSide === null) continue

    if (config.zMax > 0) {
      if (signalSide === 'short' && z > config.zMax) continue
      if (signalSide === 'long' && z < -config.zMax) continue
    }

    if (config.zReversal) {
      const zPrev = zScore(closes, closed - 1, lb)
      if (zPrev === null) continue
      if (signalSide === 'short' && z >= zPrev) continue
      if (signalSide === 'long' && z <= zPrev) continue
    }

    if (config.zFastMin > 0) {
      const zFast = zScore(closes, closed, lbF)
      if (zFast === null) continue
      if (signalSide === 'short' && zFast < config.zFastMin) continue
      if (signalSide === 'long' && zFast > -config.zFastMin) continue
    }

    if (config.rankConfirm > 0) {
      const rank = percentileRank(closes, closed, lb)
      if (rank === null) continue
      if (signalSide === 'short' && rank < config.rankConfirm) continue
      if (signalSide === 'long' && rank > 1.0 - config.rankConfirm) continue
    }

    entryBar[i] = true
    side[i] = signalSide
  }

  return { entryBar, side }
}

export function resolveBlendFade2AtOpen(
  index: number,
  candles: ChartCandle[],
  config: BlendFade2Config = DEFAULT_BLEND_FADE2_CONFIG,
  nextBarOpenTimeMs?: number,
): MarketTrend | null {
  if (index < 0) return null

  if (index < candles.length) {
    const signals = generateBlendFade2Signals(candles, config)
    return signals.entryBar[index] ? signals.side[index] : null
  }

  if (nextBarOpenTimeMs === undefined || candles.length === 0) return null

  const anchor = candles[candles.length - 1]
  const extended: ChartCandle[] = [
    ...candles,
    {
      time: nextBarOpenTimeMs,
      open: anchor.close,
      high: anchor.close,
      low: anchor.close,
      close: anchor.close,
    },
  ]
  const signals = generateBlendFade2Signals(extended, config)
  return signals.entryBar[index] ? signals.side[index] : null
}

File: src/PolyTrader.Core/Strategy/BlendFade2Signals.cs
using PolyTrader.Core.Models;

namespace PolyTrader.Core.Strategy;

public sealed class BlendFade2SignalArrays
{
    public IReadOnlyList<bool> EntryBar { get; init; } = [];
    public IReadOnlyList<MarketTrend?> Side { get; init; } = [];
}

/// <summary>
/// Blend fade 2 signal generator (1:1 with trading-cursor-models strategies/blend_fade2/signals.py).
/// Decision at open[i] using only bars [0..i-1] and open[i] timestamp for session gate.
/// </summary>
public static class BlendFade2Signals
{
    public static BlendFade2SignalArrays Generate(
        IReadOnlyList<ChartCandle> candles,
        BlendFade2Config? config = null)
    {
        var cfg = config ?? BlendFade2Config.Default;
        var n = candles.Count;
        var entry = new bool[n];
        var side = new MarketTrend?[n];

        var closes = candles.Select(c => c.Close).ToArray();
        var openTimes = candles.Select(c => c.Time).ToArray();

        var lb = cfg.Lookback;
        var lbF = cfg.LookbackFast;
        var zTh = cfg.ZThreshold;

        for (var i = 1; i < n; i++)
        {
            if (!SessionOk(openTimes[i], cfg))
            {
                continue;
            }

            var closed = i - 1;
            if (closed < Math.Max(lb, lbF) + 1)
            {
                continue;
            }

            var z = ZScore(closes, closed, lb);
            if (z is null)
            {
                continue;
            }

            if (cfg.MinRangePct > 0 && closed >= lb)
            {
                var windowStart = closed - lb;
                var refPrice = closes[windowStart];
                if (refPrice > 0)
                {
                    var windowMax = closes[windowStart];
                    var windowMin = closes[windowStart];
                    for (var j = windowStart; j <= closed; j++)
                    {
                        windowMax = Math.Max(windowMax, closes[j]);
                        windowMin = Math.Min(windowMin, closes[j]);
                    }

                    var move = (windowMax - windowMin) / refPrice;
                    if (move < cfg.MinRangePct)
                    {
                        continue;
                    }
                }
            }

            string? signalSide = null;
            if (z > zTh)
            {
                signalSide = "short";
            }
            else if (z < -zTh)
            {
                signalSide = "long";
            }

            if (signalSide is null)
            {
                continue;
            }

            if (cfg.ZMax > 0)
            {
                if (signalSide == "short" && z > cfg.ZMax)
                {
                    continue;
                }

                if (signalSide == "long" && z < -cfg.ZMax)
                {
                    continue;
                }
            }

            if (cfg.ZReversal)
            {
                var zPrev = ZScore(closes, closed - 1, lb);
                if (zPrev is null)
                {
                    continue;
                }

                if (signalSide == "short" && z >= zPrev)
                {
                    continue;
                }

                if (signalSide == "long" && z <= zPrev)
                {
                    continue;
                }
            }

            if (cfg.ZFastMin > 0)
            {
                var zFast = ZScore(closes, closed, lbF);
                if (zFast is null)
                {
                    continue;
                }

                if (signalSide == "short" && zFast < cfg.ZFastMin)
                {
                    continue;
                }

                if (signalSide == "long" && zFast > -cfg.ZFastMin)
                {
                    continue;
                }
            }

            if (cfg.RankConfirm > 0)
            {
                var rank = PercentileRank(closes, closed, lb);
                if (rank is null)
                {
                    continue;
                }

                if (signalSide == "short" && rank < cfg.RankConfirm)
                {
                    continue;
                }

                if (signalSide == "long" && rank > 1.0 - cfg.RankConfirm)
                {
                    continue;
                }
            }

            entry[i] = true;
            side[i] = signalSide == "long" ? MarketTrend.Long : MarketTrend.Short;
        }

        return new BlendFade2SignalArrays
        {
            EntryBar = entry,
            Side = side,
        };
    }

    public static MarketTrend? ResolveAtOpen(
        int index,
        IReadOnlyList<ChartCandle> candles,
        BlendFade2Config? config = null,
        long? nextBarOpenTimeMs = null)
    {
        if (index < 0)
        {
            return null;
        }

        if (index < candles.Count)
        {
            var signals = Generate(candles, config);
            return signals.EntryBar[index] ? signals.Side[index] : null;
        }

        if (nextBarOpenTimeMs is null || candles.Count == 0)
        {
            return null;
        }

        var anchor = candles[^1];
        var extended = candles.ToList();
        extended.Add(new ChartCandle
        {
            Time = nextBarOpenTimeMs.Value,
            Open = anchor.Close,
            High = anchor.Close,
            Low = anchor.Close,
            Close = anchor.Close,
        });

        var nextSignals = Generate(extended, config);
        return nextSignals.EntryBar[index] ? nextSignals.Side[index] : null;
    }

    private static double? ZScore(double[] values, int endIdx, int lookback)
    {
        var start = endIdx - lookback;
        if (start < 0)
        {
            return null;
        }

        double sum = 0;
        for (var i = start; i < endIdx; i++)
        {
            sum += values[i];
        }

        var mu = sum / lookback;
        double varSum = 0;
        for (var i = start; i < endIdx; i++)
        {
            var d = values[i] - mu;
            varSum += d * d;
        }

        var std = Math.Sqrt(varSum / lookback);
        if (std <= 0)
        {
            return null;
        }

        return (values[endIdx] - mu) / std;
    }

    private static double? PercentileRank(double[] closes, int endIdx, int lookback)
    {
        var start = endIdx - lookback;
        if (start < 0)
        {
            return null;
        }

        var lo = closes[start];
        var hi = closes[start];
        for (var i = start; i <= endIdx; i++)
        {
            lo = Math.Min(lo, closes[i]);
            hi = Math.Max(hi, closes[i]);
        }

        var span = hi - lo;
        if (span <= 0)
        {
            return null;
        }

        return (closes[endIdx] - lo) / span;
    }

    private static bool SessionOk(long openTimeUnix, BlendFade2Config cfg)
    {
        var start = cfg.SessionUtcStart;
        var end = cfg.SessionUtcEnd;
        if (start is null || end is null)
        {
            return true;
        }

        var hour = openTimeUnix > 1_000_000_000_000L
            ? (int)((openTimeUnix / 1000 / 3600) % 24)
            : (int)((openTimeUnix / 3600) % 24);

        if (start <= end)
        {
            return hour >= start && hour < end;
        }

        return hour >= start || hour < end;
    }
}
